fusequery/query/src/optimizers/optimizer_statistics_exact.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020-2021 The Datafuse Authors.
+//
+// SPDX-License-Identifier: Apache-2.0.
+
+use common_datavalues::DataValue;
+use common_exception::Result;
+use common_planners::AggregatorPartialPlan;
+use common_planners::Expression;
+use common_planners::ExpressionPlan;
+use common_planners::PlanBuilder;
+use common_planners::PlanNode;
+use common_planners::PlanRewriter;
+
+use crate::optimizers::Optimizer;
+use crate::sessions::FuseQueryContextRef;
+
+struct StatisticsExactImpl<'a> {
+    ctx: &'a FuseQueryContextRef,
+}
+
+pub struct StatisticsExactOptimizer {
+    ctx: FuseQueryContextRef,
+}
+
+impl<'plan> PlanRewriter<'plan> for StatisticsExactImpl<'_> {
+    fn rewrite_aggregate_partial(
+        &mut self,
+        plan: &'plan AggregatorPartialPlan,
+    ) -> Result<PlanNode> {
+        let new_plan = match (
```

Comment:
What happens when execute following queries?
```sql
SELECT COUNT() FROM numbers(1000) WHERE number = 500;
SELECT COUNT() FROM numbers(1000) HAVING number = 500;
```

Reply:
Nothing will happen, we will only try to match:

```
    AggregatorPartial: groupBy=[[]], aggr=[[COUNT(whatever)]]
      Expression: whatever
         ReadDataSource: whatever with exact statistics
```

Any filter will disable the optimization

Reply:
What'more, `count(0)` is required, because `count()` wil missing a (Before GroupBy) Expression.

We can write another optimizer to omit any arguments in count, and make this more generic

Reply:
It' ok.  BTW how about checking the nullable for the first expression?

http://sqlfiddle.com/#!9/34d4ba/5

Reply:
Good catch, so we can only omit count(args) with non-null fields. We should check the aggr.rule

Or for simple, we can only optimize for `count(literal)` and `count()`

Reply:
I prefer just to optimize `count(literal)` and `count()` is ok.  

People who use count(other_expressions) may do not like the optimization, like `select count(sleep(2)) from numbers(10)`

Reply:
Fixed